File: lib/karabinex/karabiner_config.ex
defmodule Karabinex.KarabinerConfig do
  @type json_map :: %{optional(String.t()) => term()}
  @type rule :: json_map()
  @type rules :: [rule()]

  @karabiner_config_path "~/.config/karabiner/karabiner.json"
  @karabinex_rules_path "karabinex.json"

  @spec update() :: :ok
  def update do
    rules = read_rules!(@karabinex_rules_path)
    config_path = Path.expand(@karabiner_config_path)
    config = read_json!(config_path)
    updated = replace_rules_in_config(config, rules)

    if updated == config do
      :ok
    else
      write_json!(config_path, updated)
    end
  end

  @spec read_rules!(Path.t()) :: rules()
  defp read_rules!(path) do
    case read_json!(path) do
      %{"rules" => rules} when is_list(rules) and rules != [] ->
        rules

      %{"rules" => []} ->
        raise "karabinex.json has no rules"

      _ ->
        raise "karabinex.json missing rules"
    end
  end

  @spec read_json!(Path.t()) :: json_map()
  defp read_json!(path) do
    path
    |> File.read!()
    |> Jason.decode!()
  end

  @spec write_json!(Path.t(), json_map()) :: :ok
  defp write_json!(path, data) do
    File.write!(path, Jason.encode!(data, pretty: true))
  end

  @spec replace_rules_in_config(json_map(), rules()) :: json_map()
  def replace_rules_in_config(%{"profiles" => profiles} = config, new_rules)
      when is_list(profiles) do
    updated_profiles = Enum.map(profiles, &replace_rules_in_profile(&1, new_rules))
    %{config | "profiles" => updated_profiles}
  end

  def replace_rules_in_config(_config, _new_rules) do
    raise "Karabiner config missing profiles"
  end

  @spec replace_rules_in_profile(json_map(), rules()) :: json_map()
  defp replace_rules_in_profile(
         %{"complex_modifications" => %{"rules" => rules} = complex} = profile,
         new_rules
       )
       when is_list(rules) do
    case replace_rules(rules, new_rules) do
      {updated_rules, true} ->
        %{profile | "complex_modifications" => %{complex | "rules" => updated_rules}}

      {_rules, false} ->
        profile
    end
  end

  defp replace_rules_in_profile(profile, _new_rules), do: profile

  @spec replace_rules([json_map()], rules()) :: {[json_map()], boolean()}
  defp replace_rules(existing_rules, new_rules) do
    descriptions =
      new_rules
      |> Enum.map(&Map.get(&1, "description"))
      |> Enum.filter(&is_binary/1)
      |> MapSet.new()

    if MapSet.size(descriptions) == 0 do
      raise "karabinex.json rules missing descriptions"
    end

    {rev, inserted?} =
      Enum.reduce(existing_rules, {[], false}, fn rule, {acc, inserted?} ->
        description = Map.get(rule, "description")

        if MapSet.member?(descriptions, description) do
          if inserted? do
            {acc, true}
          else
            {Enum.reverse(new_rules, acc), true}
          end
        else
          {[rule | acc], inserted?}
        end
      end)

    if inserted? do
      {Enum.reverse(rev), true}
    else
      {existing_rules, false}
    end
  end
end

File: src/karabinerConfig.ts
import { readFileSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";

export type JsonMap = Record<string, unknown>;
export type Rule = JsonMap;

const defaultKarabinerConfigPath = join(
  homedir(),
  ".config/karabiner/karabiner.json",
);

export function updateKarabinerConfig(
  karabinexRulesPath: string,
  configPath = defaultKarabinerConfigPath,
): void {
  const rules = readRules(karabinexRulesPath);
  const config = readJson(configPath);
  const { updated, changed } = replaceRulesInConfig(config, rules);

  if (!changed) {
    return;
  }

  writeJson(configPath, updated);
}

function readRules(path: string): Rule[] {
  const data = readJson(path);
  const rules = (data as JsonMap).rules;

  if (Array.isArray(rules) && rules.length > 0) {
    return rules as Rule[];
  }

  if (Array.isArray(rules) && rules.length === 0) {
    throw new Error("karabinex.json has no rules");
  }

  throw new Error("karabinex.json missing rules");
}

function readJson(path: string): JsonMap {
  return JSON.parse(readFileSync(path, "utf8")) as JsonMap;
}

function writeJson(path: string, data: JsonMap): void {
  writeFileSync(path, `${JSON.stringify(data, null, 2)}\n`);
}

function replaceRulesInConfig(
  config: JsonMap,
  newRules: Rule[],
): { updated: JsonMap; changed: boolean } {
  const profiles = config.profiles;
  if (!Array.isArray(profiles)) {
    throw new Error("Karabiner config missing profiles");
  }

  let changed = false;
  const updatedProfiles = profiles.map((profile) => {
    const { updated, changed: profileChanged } = replaceRulesInProfile(
      profile as JsonMap,
      newRules,
    );
    if (profileChanged) {
      changed = true;
    }
    return updated;
  });

  if (!changed) {
    return { updated: config, changed: false };
  }

  return { updated: { ...config, profiles: updatedProfiles }, changed: true };
}

function replaceRulesInProfile(
  profile: JsonMap,
  newRules: Rule[],
): { updated: JsonMap; changed: boolean } {
  const complex = profile.complex_modifications as JsonMap | undefined;
  const rules = complex?.rules;
  if (!Array.isArray(rules)) {
    return { updated: profile, changed: false };
  }

  const { updated, changed } = replaceRules(rules as Rule[], newRules);
  if (!changed) {
    return { updated: profile, changed: false };
  }

  return {
    updated: {
      ...profile,
      complex_modifications: {
        ...complex,
        rules: updated,
      },
    },
    changed: true,
  };
}

function replaceRules(
  existingRules: Rule[],
  newRules: Rule[],
): { updated: Rule[]; changed: boolean } {
  const descriptions = new Set(
    newRules
      .map((rule) => rule.description)
      .filter(
        (description): description is string => typeof description === "string",
      ),
  );

  if (descriptions.size === 0) {
    throw new Error("karabinex.json rules missing descriptions");
  }

  const updated: Rule[] = [];
  let inserted = false;

  for (const rule of existingRules) {
    const description = rule.description;
    if (typeof description === "string" && descriptions.has(description)) {
      if (!inserted) {
        updated.push(...newRules);
        inserted = true;
      }
    } else {
      updated.push(rule);
    }
  }

  if (!inserted) {
    return { updated: existingRules, changed: false };
  }

  return { updated, changed: true };
}
